test/lib/mutation_source_test.cc

```diff
@@ -2332,3 +2332,82 @@ void for_each_schema_change(std::function<void(schema_ptr, const std::vector<mut
     }
     test_mutated_schemas();
 }
+
+class mfo_printer {
+    const schema& _s;
+    const mutation_fragment_opt& _mfo;
+public:
+    mfo_printer(const schema& s, const mutation_fragment_opt& mf) : _s(s), _mfo(mf) { }
+    mfo_printer(const mfo_printer&) = delete;
+    mfo_printer(mfo_printer&&) = delete;
+
+    friend std::ostream& operator<<(std::ostream& os, const mfo_printer& p) {
+        if (p._mfo) {
+            os << mutation_fragment::printer(p._s, *p._mfo);
+        } else {
+            os << "eos";
+        }
+        return os;
+    }
+};
+
+using event = std::variant<mutation_fragment_opt, position_range>;
+using trace = std::vector<event>;
+
+class trace_printer {
+    const schema& _s;
+    const trace& _trace;
+public:
+    trace_printer(const schema& s, const trace& t) : _s(s), _trace(t) { }
+
+    friend std::ostream& operator<<(std::ostream& os, const trace_printer& p) {
+        for (auto& e: p._trace) {
+            os << "\n";
+            std::visit(make_visitor(
+                [&] (const mutation_fragment_opt& mf) { os << mfo_printer(p._s, mf); },
+                [&] (const position_range& pr) { os << "forwarding to " << pr; }
+            ), e);
+        }
+        return os;
+    }
+};
+
+static void compare_readers(const schema& s, trace& trace, flat_mutation_reader& authority, flat_mutation_reader& tested) {
+    while (auto expected = authority(db::no_timeout).get()) {
+        auto mf = tested(db::no_timeout).get();
+        if (!mf) {
+            BOOST_FAIL(format("compare_readers: expected:\n{},\nbut the reader returned eos. Trace:{}",
+                        mfo_printer(s, expected), trace_printer(s, trace)));
+        }
+
+        if (!expected->equal(s, *mf)) {
+            BOOST_FAIL(format("compare_readers: expected:\n{},\nbut got:\n{}.\nTrace:{}",
+                        mfo_printer(s, expected), mfo_printer(s, mf), trace_printer(s, trace)));
+        }
+
+        trace.push_back(std::move(expected));
+    }
+
+    auto mf = tested(db::no_timeout).get();
+    if (mf) {
+        BOOST_FAIL(format("compare_readers: expected eos, but the reader returned:\n{}.\nTrace:{}",
+                    mfo_printer(s, mf), trace_printer(s, trace)));
+    }
+    trace.push_back(std::move(mf));
```

Comment:
I think this could be much easily done with `flat_mutation_reader_assertions()`:
```
auto rd_auth = make_authority();
auto rd_tested = assert_that(make_tested());
while (auto mf_opt = rd_auth(db::no_timeout).get0(); mf_opt) {
    rd_tested.produces(*mf_opt);
}
rd_tested.produces_end_of_stream();
```

Reply:
Done